### PR TITLE
Correct branch count in bmt cli input, add hashing to generation

### PR DESCRIPTION
**Pull request by nolash**

### CODE REVIEW

File: cmd/generate-hashes/generate-hashes.go [hunk 1]
```diff
@@ -31,16 +39,36 @@ func main() {
 		os.Exit(1)
 	}
 
+	// set up hasher
+	hashFunc := func() hash.Hash {
+		return sha3.NewLegacyKeccak256()
+	}
+	hashPool := legacy.NewTreePool(hashFunc, 128, legacy.PoolSize)
```
Comment (acud) [inline cmd/generate-hashes/generate-hashes.go:46]: please comment that 128 is only true for unencrypted chunks

File: cmd/main_legacy.go [hunk 2]
```diff
@@ -47,7 +47,7 @@ func main() {
 	hashFunc := func() hash.Hash {
 		return sha3.NewLegacyKeccak256()
 	}
-	hashPool := legacy.NewTreePool(hashFunc, 32, legacy.PoolSize)
```
Comment (acud) [inline cmd/main_legacy.go:50]: nice one

**Review by acud — APPROVED** [review]

File: cmd/generate-hashes/generate-hashes.go [hunk 3]
```diff
@@ -21,8 +25,12 @@ func main() {
...
 		os.Exit(1)
 	}
-	outputDir := os.Args[1]
-	err := os.Mkdir(outputDir, 0755)
+	outputDir, err := filepath.Abs(os.Args[1])
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Invalid input: %s", err)
+		os.Exit(1)
+	}
+	err = os.Mkdir(outputDir, 0755)
```
Comment (janos) [inline cmd/generate-hashes/generate-hashes.go]: I would leave here mask 0777 as this is making not possible for user to have more permissive access on group and other. It represents permissions "before umask".

File: cmd/generate-hashes/generate-hashes.go [hunk 4]
```diff
@@ -31,16 +39,36 @@ func main() {
...
+			os.Exit(1)
+		}
+		bmtHash.Reset()
+		_, err = bmtHash.Write(b)
+		sum := bmtHash.Sum(nil)
+		if err != nil {
+			fmt.Fprint(os.Stderr, err.Error())
+			os.Exit(1)
+		}
+		err = ioutil.WriteFile(filename, sum, 0644)
```
Comment (janos) [inline cmd/generate-hashes/generate-hashes.go]: My suggestion is to use 0666 file mode, as well as a few lines below, which I probably missed to see in previous pr.

**Review by janos — APPROVED** [review]
My comments are not crucial for this pr, they can be addressed later in some future pr.

File: cmd/generate-hashes/generate-hashes.go (hunk 1 above)
Comment (nolash) [reply cmd/generate-hashes/generate-hashes.go:46]: Well, while that's true from a functional perspective, it's not quite true here. The chunk is still 4096 bytes and segment size is determined by the hasher. There's no other way to tell the BMT that it's anything else than `32*128`, unless we mess with the underlying hasher to also report `Size()` to be `64`. Actually the idea was the _new famous_ filehasher can accommodate for this.

In fact, I believe the bmt should be agnostic to all of this. It should care about hashing digest-size segments together and that's it.

File: cmd/generate-hashes/generate-hashes.go (hunk 4 above)
Comment (nolash) [reply cmd/generate-hashes/generate-hashes.go]: :guitar:

**Pull request merged**